scalameta/semantic/shared/src/main/scala/scala/meta/internal/semantic/Schema.scala

```diff
@@ -0,0 +1,97 @@
+package scala.meta
+package internal
+package semantic
+
+import java.net.URI
+import org.scalameta.data._
+import scala.{Seq => _}
+import scala.collection.immutable.Seq
+import scala.meta.io._
+import scala.meta.internal.semantic.{vfs => v}
+import scala.meta.internal.semantic.{schema => s}
+import scala.meta.{semantic => m}
+import scala.meta.inputs.{Input => mInput, Position => mPosition, Point => mPoint}
+
+package object schema {
+  @data class Database(entries: Seq[Attributes]) {
+    def toVfs(classpath: Classpath, sourcepath: Sourcepath): v.Database = {
+      if (classpath.shallow.isEmpty) sys.error("can't save semanticdb to an empty classpath")
+      val ventries = entries.map(entry => {
+        // TODO: Would it make sense to support multiclasspaths?
+        // One use-case for this would be in-place updates of semanticdb files.
+        val base = AbsolutePath(classpath.shallow.head)
+        val scalaNameOpt = sourcepath.relativize(AbsolutePath(entry.path).toURI)
+        val scalaName = scalaNameOpt.getOrElse(sys.error(s"can't find ${entry.path} in $sourcepath"))
+        val semanticdbName = v.Paths.scalaToSemanticdb(scalaName)
+        val fragment = Fragment(base, semanticdbName)
+        v.Entry.InMemory(fragment, entry.toByteArray)
+      })
+      v.Database(ventries)
+    }
+
+    def toMeta: m.Database = {
```

Comment:
I've rewritten serialization/deserialization logic from typeclasses to plain old Scala code, because I found that type signatures of typeclass instances really got in the way.